lib/navigation/preprocess_imu.cpp

```diff
@@ -9,48 +15,158 @@ ImuPreprocessor::ImuPreprocessor()
 
 core::ImuData ImuPreprocessor::processData(const core::RawImuData raw_imu_data)
 {
-  /*
-  TODOLater: implement
-  basic plan:
-  - call detectOutliers. Return of that function is return of this function
-  - call checkReliable
-
-  */
-  return {0, 0, 0, 0};
+  core::ImuData clean_accelerometer_data;
+  clean_accelerometer_data = detectOutliers(raw_imu_data);
+  checkReliable();
+  return clean_accelerometer_data;
 }
 
 core::ImuData ImuPreprocessor::detectOutliers(const core::RawImuData imu_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  outlier_imus_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  core::ImuData clean_accelerometer_data;
+  // TODOLater : change magnitude for directions (hardware configuration?)
+  core::Float magnitude;
+  for (std::size_t i = 0; i < core::kNumImus; ++i) {
+    magnitude = 0;
+    for (std::size_t j = 0; j < 3; ++j) {
+      magnitude += std::pow(imu_data.at(i).at(j), 2);
+    }
+    clean_accelerometer_data.at(i) = std::sqrt(magnitude);
+  }
+  Quartiles quartiles     = getOutlierThresholds(clean_accelerometer_data);
```

Comment:
@aryansharma19 Look at the quartiles struct. This is a neater implementation and will work for encoders data too.